File: lab1/huffman.ml
type bit = Zero | One
let bit_from_int = function
    | 0 -> Zero
    | _ -> One
let int_from_bit = function
    | Zero -> 0
    | One -> 1

type 'a tree = Leaf of 'a | Node of 'a tree * 'a tree

let get_weights xs =
    let mp = Hashtbl.create 256 in
    for i = 0 to Array.length xs - 1 do
        match Hashtbl.find_opt mp xs.(i) with
            | None -> Hashtbl.add mp xs.(i) 1
            | Some v -> Hashtbl.replace mp xs.(i) (v + 1)
    done;
    let l = Hashtbl.fold (fun k v acc -> (k, v) :: acc) mp [] in
    let l' = List.sort (fun (k, v) (k', v') -> Int.min v v') l in
    List.mapi (fun i (x, _) -> (x, i)) l';;

(* TODO: Избавиться от Option.get *)
let rec from_queue queue =
    if Priority_queue.is_single queue
    then
        let (_, value, _) = Option.get (Priority_queue.extract queue)
        in value
    else
        let (p1, v1, queue) = Option.get (Priority_queue.extract queue) in
        let (p2, v2, queue) = Option.get (Priority_queue.extract queue) in
        from_queue (Priority_queue.insert queue (p1 + p2) (Node (v1, v2)));;

let get_code = function
    | Leaf value -> Hashtbl.of_seq (List.to_seq [(value, [One])])
    | tree ->
        let rec aux acc run = function
            | Leaf value -> (value, List.rev run) :: acc
            | Node (left, right) ->
                (aux acc (Zero :: run) left) @ (aux acc (One :: run) right)
        in Hashtbl.of_seq (List.to_seq (aux [] [] tree));;

type metadata = {
    weights : (char * int) list;
    tree : char tree;
    code : (char, bit list) Hashtbl.t;
    remainder : char;
};;

let dump_metadata metadata =
    let wlen = List.length metadata.weights in
    let weights = List.fold_right
        (fun (character, weight) acc -> acc @ [character; Char.chr weight])
        metadata.weights
        [] in
    metadata.remainder :: Char.chr (wlen mod 256) :: weights

let load_metadata (data : char array) =
    let remainder = data.(0) in
    let wlen = int_of_char data.(1) in
    let wlen = if wlen == 0 then 256 else wlen in
    let weights = ref [] in
    for i = 0 to wlen - 1 do
        let character = data.(2 + i * 2) in
        let weight = int_of_char data.(2 + i * 2 + 1) in
        weights := !weights @ [(character, weight)]
    done;

    let queue = List.fold_right
        (fun (character, weight) acc ->
            Priority_queue.insert acc weight (Leaf character))
        (List.rev !weights)
        Priority_queue.Empty in
    let tree = from_queue queue in
    let code = get_code tree in
    { remainder = remainder; weights = !weights; tree = tree; code = code }

let skip_metadata data =
    let size = Array.length data in
    let wlen = int_of_char data.(1) in
    Array.sub data (2 + wlen * 2) (size - wlen * 2 - 2);;

let compute_metadata data =
    let weights = get_weights data in
    let queue = List.fold_right
        (fun (character, weight) acc ->
            Priority_queue.insert acc weight (Leaf character))
        weights
        Priority_queue.Empty in
    let tree = from_queue queue in
    let code = get_code tree in
    { remainder = char_of_int 0;
      weights = weights; tree = tree;
      code = code };;

class bit_compressor =
    object (self)
        val mutable buffer : bit list = []
        val mutable remainder : char = char_of_int 0
        val mutable result : char list = []

        method private output_char byte =
            result <- byte :: result
        method private dump_byte =
            let byte = ref 0 in
                for i = 0 to min 7 (List.length buffer - 1) do
                    let bit = int_from_bit (List.hd buffer) in
                    byte := Int.logor !byte (Int.shift_left bit i);
                    buffer <- List.tl buffer
                done;
            self#output_char (char_of_int !byte)

        method write_metadata metadata =
            let dumped = dump_metadata metadata in
            List.iter self#output_char dumped
        method write_bit bit =
            buffer <- buffer @ [bit];
            if List.length buffer == 8 then self#dump_byte
        method write_bits = List.iter self#write_bit
        method close =
            let remainder' = (8 - List.length buffer) mod 8 in begin
                if remainder' != 0 then self#dump_byte;
                remainder <- char_of_int remainder';
            end
        method get_result =
            let arr = Array.of_seq (List.to_seq (List.rev result)) in
            arr.(0) <- remainder; arr
    end;;

class bit_decompressor (archive : char array) =
    object (self)
        val metadata = load_metadata archive
        method metadata = metadata
        val data = skip_metadata archive
        val mutable buffer : int list = []
        val mutable ptr = 0

        method private get_opt arr idx =
            try Some arr.(idx)
            with Invalid_argument _ -> None

        method private read_byte =
            let bits = match self#get_opt data ptr with
                | None -> []
                | Some byte ->
                    let num = int_of_char byte in
                    let rec aux acc = function
                        | 8 -> List.rev acc
                        | n -> let shifted = Int.shift_right num n in
                               let bit = Int.logand 1 shifted in
                               aux (bit :: acc) (n + 1)
                    in aux [] 0
            in begin
                let bits = if ptr == Array.length data - 1 then
                    List.rev (List.of_seq (
                        Seq.drop (int_of_char metadata.remainder)
                                 (List.to_seq (List.rev bits))))
                else bits
                in begin
                    buffer <- buffer @ bits;
                    ptr <- ptr + 1
                end
            end

        method read_bit =
            if List.length buffer == 0 then self#read_byte;
            match buffer with
                | [] -> None
                | hd :: tl -> (buffer <- tl; Some (bit_from_int hd))
    end;;

let compress bin_data =
    let metadata = compute_metadata bin_data in
    let stream = new bit_compressor in
    stream#write_metadata metadata;
    Array.iter (fun c ->
        stream#write_bits (Hashtbl.find metadata.code c)) bin_data;
    stream#close;
    stream#get_result;;

let decompress data =
    let decompressor = new bit_decompressor data in
    let tree = decompressor#metadata.tree in
    match tree with
    | Leaf value ->
        let rec aux acc =
            match decompressor#read_bit with
            | None -> List.rev acc
            | Some _ -> aux (value :: acc)
        in Array.of_seq (List.to_seq (aux []))
    | tree ->
        let rec aux acc = function
            | Leaf value -> aux (value :: acc) tree
            | Node (left, right) ->
                match decompressor#read_bit with
                    | None -> List.rev acc
                    | Some Zero -> aux acc left
                    | Some One -> aux acc right
        in Array.of_seq (List.to_seq (aux [] tree));;

File: lab1-rs/src/huffman.rs
use crate::priority_queue::PriorityQueue;
use std::collections::HashMap;

#[derive(Clone, Copy, Debug)]
enum Bit {
    Zero,
    One,
}
impl Bit {
    fn from_u8(x: u8) -> Bit {
        if x == 0 {
            Bit::Zero
        } else {
            Bit::One
        }
    }

    fn to_u8(x: &Bit) -> u8 {
        match *x {
            Bit::Zero => 0,
            Bit::One => 1,
        }
    }
}

#[derive(Debug)]
enum HuffmanTree<T>
where
    T: std::cmp::Eq + std::hash::Hash + Copy,
{
    Leaf(T),
    Node(Box<HuffmanTree<T>>, Box<HuffmanTree<T>>),
}

impl<T: std::cmp::Eq + std::hash::Hash + Copy> HuffmanTree<T> {
    fn from_queue(mut queue: PriorityQueue<T>) -> Self {
        let mut huf_queue = PriorityQueue::new();
        while let Some(p) = queue.pop() {
            huf_queue.insert(p.weight, Box::new(HuffmanTree::Leaf(p.value)));
        }
        assert!(huf_queue.len() > 0);
        while huf_queue.len() > 1 {
            let v1 = huf_queue.pop().unwrap();
            let v2 = huf_queue.pop().unwrap();
            huf_queue.insert(
                v1.weight + v2.weight,
                Box::new(HuffmanTree::Node(v1.value, v2.value)),
            );
        }
        return *huf_queue.pop().unwrap().value;
    }

    fn dfs(tree: &HuffmanTree<T>, mut acc: Vec<Bit>, code: &mut HashMap<T, Vec<Bit>>) -> Vec<Bit> {
        match tree {
            HuffmanTree::Leaf(val) => {
                code.insert(*val, acc.clone());
            }
            HuffmanTree::Node(left, right) => {
                acc.push(Bit::Zero);
                acc = HuffmanTree::dfs(left, acc, code);
                acc.pop();
                acc.push(Bit::One);
                acc = HuffmanTree::dfs(right, acc, code);
                acc.pop();
            }
        };
        return acc;
    }

    fn get_code(&self) -> HashMap<T, Vec<Bit>> {
        let mut code = HashMap::new();
        if let HuffmanTree::Leaf(val) = &self {
            code.insert(*val, vec![Bit::One]);
        } else {
            HuffmanTree::dfs(self, Vec::new(), &mut code);
        }
        return code;
    }
}

fn get_weights(data: &Vec<u8>) -> HashMap<u8, u32> {
    let mut freq = HashMap::new();
    data.iter().for_each(|byte| {
        *freq.entry(*byte).or_insert(1) += 1;
    });
    let mut freq: Vec<(u8, u32)> = freq.drain().collect();
    freq.sort();
    let mut weights = HashMap::new();
    freq.iter().enumerate().for_each(|(i, (byte, _))| {
        weights.insert(*byte, i as u32);
    });
    return weights;
}

#[derive(Debug)]
struct Metadata {
    weights: HashMap<u8, u32>,
    tree: HuffmanTree<u8>,
    code: HashMap<u8, Vec<Bit>>,
    remainder: u8,
}

impl Metadata {
    fn compute(data: &Vec<u8>) -> Self {
        let weights = get_weights(data);
        let mut queue = PriorityQueue::new();
        for (value, priority) in weights.iter() {
            queue.insert(*priority, *value);
        }
        let tree = HuffmanTree::from_queue(queue);
        let code = tree.get_code();

        return Self {
            weights,
            tree,
            code,
            remainder: 0,
        };
    }

    fn load(data: &Vec<u8>) -> Self {
        let remainder = data[0];
        let dict_len = data[1] as usize + 1;
        let mut weights = HashMap::new();
        for i in 0..dict_len {
            weights.insert(data[2 + 2 * i], data[2 + 2 * i + 1] as u32);
        }

        let mut queue = PriorityQueue::new();
        for (value, priority) in weights.iter() {
            queue.insert(*priority, *value);
        }
        let tree = HuffmanTree::from_queue(queue);
        let code = tree.get_code();

        return Self {
            weights,
            tree,
            code,
            remainder,
        };
    }

    fn dump(&self) -> Vec<u8> {
        let mut result = Vec::new();

        result.push(self.remainder);
        result.push((self.weights.len() - 1) as u8);
        for (byte, weight) in &self.weights {
            result.push(*byte);
            result.push(*weight as u8);
        }

        return result;
    }
}

struct BitWriter {
    buffer: Vec<Bit>,
    remainder: u8,
    result: Vec<u8>,
}

impl BitWriter {
    fn new() -> Self {
        Self {
            buffer: Vec::new(),
            remainder: 0,
            result: Vec::new(),
        }
    }

    fn dump_byte(&mut self) {
        let mut byte = 0_u8;
        self.buffer
            .iter()
            .map(Bit::to_u8)
            .enumerate()
            .for_each(|(i, bit)| byte |= bit << i);
        self.result.push(byte);
        self.buffer.clear();
    }

    fn write_bit(&mut self, bit: Bit) {
        self.buffer.push(bit);
        if self.buffer.len() == 8 {
            self.dump_byte();
        }
    }

    fn write_bits(&mut self, bits: &Vec<Bit>) {
        bits.iter().for_each(|bit| self.write_bit(*bit));
    }

    fn finish(&mut self) {
        let remainder = 8 - self.buffer.len() as u8 % 8;
        if remainder != 0 {
            self.dump_byte();
        }
        self.remainder = remainder;
    }
}

pub fn compress(data: &Vec<u8>) -> Vec<u8> {
    let mut result = Vec::new();
    let mut metadata = Metadata::compute(data);

    let mut writer = BitWriter::new();
    data.iter().for_each(|byte| {
        let bits = metadata.code.get(byte).unwrap();
        writer.write_bits(&bits);
    });
    writer.finish();
    metadata.remainder = writer.remainder;

    let md_dump = metadata.dump();
    md_dump.iter().for_each(|byte| result.push(*byte));
    writer.result.iter().for_each(|byte| result.push(*byte));

    return result;
}

struct BitReader<'a> {
    data: &'a Vec<u8>,
    metadata: &'a Metadata,
    buffer: Vec<Bit>,
    ptr: usize,
}

impl<'a> BitReader<'a> {
    fn new(data: &'a Vec<u8>, metadata: &'a Metadata) -> Self {
        Self {
            data,
            metadata,
            buffer: Vec::new(),
            ptr: 0,
        }
    }

    fn read_byte(&mut self) {
        if let Some(byte) = self.data.get(self.ptr) {
            for i in 0..=7 {
                let bit = byte & (1 << i);
                self.buffer.push(Bit::from_u8(bit));
            }
            if self.ptr == self.data.len() - 1 {
                for _ in 0..self.metadata.remainder {
                    self.buffer.pop();
                }
            }
            self.buffer.reverse();
            self.ptr += 1;
        }
    }

    fn read_bit(&mut self) -> Option<Bit> {
        if self.buffer.len() == 0 {
            self.read_byte();
        }
        self.buffer.pop()
    }
}

pub fn decompress(archive: &Vec<u8>) -> Vec<u8> {
    let mut result = Vec::new();
    let metadata = Metadata::load(archive);

    let data = archive[2 + metadata.weights.len() * 2..].to_vec();
    let mut reader = BitReader::new(&data, &metadata);

    let mut state = &metadata.tree;
    while let Some(bit) = reader.read_bit() {
        if let HuffmanTree::Node(left, right) = state {
            match bit {
                Bit::Zero => state = left,
                Bit::One => state = right,
            }
        }

        if let HuffmanTree::Leaf(byte) = state {
            result.push(*byte);
            state = &metadata.tree;
        }
    }

    return result;
}
